[READY] Goon Genetics

Time-Green opened the pull request:
### GENETICS

- The random hexadecimal rng game has been replaced with gene sequencing from goon.

- Adds mutation activators and mutators

- You can now store mutations

- Everyone now has their own set of unique mutations

- Limited mutations per person to 8 (including one always being monkey)

- Adds race specific mutations (See fire breathing for lizads)

- You can inspect discovered mutations, undiscovered mutations use an alias to recognize them by

- Adds a sequence analyzer. Can be used to scan someones genes sequence. Useful for determing what mutations they can safely have and or collecting data for very difficult sequences

- Adds mutation combining. It's currently only RADIOACTIVE + STRONG = HULK  (So yes you will now need 2 mutations for to get hulk)

- Adds several other mutations. Telepathy, firebreath, glowy, radioactive and strength

:cl: Time-Green
add: Goon genetics!
add: More mutations! Fire breath for lizards! Radioactive! Telepathy! Glowy! Strength, though its cosmetic and should be combined with radioactivity instead! Fiery sweat! 
add: Adds void magnet mutation by @tralezab !
/:cl:

Quick guide on how to have fun with mutations:

Every mutation consists of a sequence. It's the AT GC stuff. A goes with T and G goes with C. An active mutation has a correct sequence, like here:
![image](https://user-images.githubusercontent.com/7501474/48154715-99b6d480-e2c9-11e8-99fa-336716e1909e.png)

First we want to disable this mutation, so we can move on with the others. Do this by making it faulty.
![image](https://user-images.githubusercontent.com/7501474/48154882-f31f0380-e2c9-11e8-9be7-65d6c40374c9.png)
We replaced the fourth A with an X, making it a wrong sequence and thus disabling the mutation.

Now we can start with the real stuff. We'll select another mutation.
![image](https://user-images.githubusercontent.com/7501474/48154970-311c2780-e2ca-11e8-8d11-9ed04dbcf976.png)
There's four X's, making it wrong. We can instantly fix two X's. The first pair with GX should be GC. The third pair with XA should be TA.

Now our mutation should look like this:
![image](https://user-images.githubusercontent.com/7501474/48155062-72143c00-e2ca-11e8-9f9e-595f09aba776.png)
That just leaves a pair of X's. There's only four possibilities: AT, TA, GC and CG

![image](https://user-images.githubusercontent.com/7501474/48155210-c3bcc680-e2ca-11e8-9854-599d1fde5c2c.png)
It was CG. The mutation is now active and can be properly toyed with. 
Note: With two X pairs, there is still only 16 possibilities. If there are even more, you might consider peaking at someone else with the same mutation to see if they got that pair. (I recommend you do this with the sequence analyzer in your pocket, wich you can click on the DNA console to update it with the latest mutation names)

If we saved the mutation, it goes to our storage (Mutations)
![image](https://user-images.githubusercontent.com/7501474/48155421-3ded4b00-e2cb-11e8-8e56-2058e7b50b11.png)

Here we can do a bunch of stuff with it. 
Mutation name: Shows you more info
![image](https://user-images.githubusercontent.com/7501474/48155551-81e05000-e2cb-11e8-97a1-10f5e99f501e.png)
Export: Export the mutation to a disk for storage
Delete: Removes the mutation from your storage
Combine: First click, selects the mutation. Second click, attempts to merge it with another mutation to create a new one. Currently only works with Strength and Radioactive

Other notes about UI:
Removed rejuvenators. Get some from the chemist or use antitoxins. Note that most radiation damage has been reduced by alot.
Scramble DNA: Don't like the your subjects mutations and can't get a new one? Absolutely fucking radiation blast them and generate a new set of mutations
Disk: Disk interaction. Import, export and disk ejecting fall under this
Genetic Sequencer: Where you will be spending most of your time
Unique Identifiers: All identity mutation stuff has been moved to there. It still features buffers and the UE gene pulser for all your identity stealing needs.

Activators! Already got a mutation and want to give it to someone with it in their potential? Ram this thing into them to safely give them that mutation.
Mutators! Want to give someone a mutation that isn't listed on them? Use these to instantly give it to them for nothing*! *(actually violent humiliation and death)

Featuring more mutations!
Firebreath!
Telepathy!
Glowy!
Radioactive!
Strength! (Doesnt do anything)
Fiery sweat! 
Void magnet! Made by @tralezab 
And even some old ones!
Also disabled the unstable dna mutation, since it doesn’t fit well with goon genetics

End of guide 

- [X] Make mutations random for every person
- [X] New html
- [X] Make mutations stable and working
- [x] Add mutation disks
- [X] Mutation activators
- [X] Mutation injectors
- [X] More mutations
- [x] Even more mutations
- [x] Have genetic instability do something
- [x] Add some form of viewing your own DNA for discovered mutations
- [X] Alot of testing
- [x] Add the goon gene minigame

I will also be testing everything thoroughly and tweaking some stuff.

Another PR:
Chromosomes
Even more than more mutations! (Probably by someone else since I'll be prioritizing chromosomes)

Time-Green committed:
severs mutations from glob

Time-Green committed:
Day 2: Mutations are actually functional

Time-Green committed:
Day 3: DNA consoles now work with the new system

Time-Green committed:
Day 4: UI additions, mutation info and experimental mutation storage

Time-Green committed:
Day 5: Adds activators, bugfixes and probably more

It's actually PR'able now, but I still need to properly do the order of the mutation index, add more mutations, add mutation disks and make sure everyone has certain genes like monkey

Time-Green committed:
day 6: just cleans up some code

Time-Green committed:
Day 7: its fucking done

now just gotta make up some cool mutations, some downsides to bistructural enzyme instability and test the living fuck out of it to make sure i didnt fuck up any mutation related stuff like wizard mutate and timers

Time-Green committed:
night 7: added radiation and spent 3 hours debugging something that i still dont understand but is fixed now

end me

Time-Green committed:
Day 8: adds a bunch of cool new mutations, spell support and probably fixed some more bugs

Time-Green committed:
Day 9: Reworks alot of code, adds more comments and makes everything more consistent.

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into spicy-genes

Time-Green committed:
Fixes conflicts

im really worried that these are the only conflicts

Time-Green committed:
Removes some comments for myself and an ugly span

Tlaltecuhtli commented:
can we get race specific mutations too?

Time-Green commented:
Read my PR. I already added them. Lizards can now unlock the ability to breath fire

Time-Green commented:
I already added them. Lizards can now unlock the ability to breath fire. More will have to be added later

Time-Green committed:
Activates some old mutations

k-k-karma commented:
HOOOOOOOOLY SHIT

anconfuzedrock commented:
Inb4 to give genetics a way to experiment on lizards you can now turn wags into a humanoid

Really though this seems rad, genetics was the first job I ever played but it was so arbitrary and boring that I could never get into it. 
I assume hulk is human only as it already was?

Time-Green commented:
Yeah, still only works on humans, though other species can still have it.
I might replace monkey with hulk as the specified human mutation and just give everyone monkey

Arianya commented:
Does the glowy mutation have an interaction with lighteater from the nightmare? Something like disabling the mutation temporarily.

ExcessiveUseOfCobblestone commented:
> To counteract the fact that genetics just got 100x as hard, you can now see who has wich mutations if they have already been discovered. This kinda ruins having to check it yourself wich was nice

It doesnt matter tho because their dna doesn't match with yours?

Nabski89 commented:
What is genetic stability?

bobbahbrown commented on code/datums/mutations/radioactive.dm:
```diff
@@ -0,0 +1,17 @@
+/datum/mutation/human/radioactive
+	name = "Radioactivity"
+	desc = "A volatile mutation that causes the host to sent out deadly beta radiation. This affects both the hosts and his surroundings."
```
the description should not use "his", try "their" or "his or her"

Time-Green replied:
I was planning to rewrite them at some point since they were done in a pinch, I'll go ahead and change that when I do

Time-Green commented:
@ExcessiveUseOfCobblestone 
> > To counteract the fact that genetics just got 100x as hard, you can now see who has wich mutations if they have already been discovered. This kinda ruins having to check it yourself wich was nice
> 
> It doesnt matter tho because their dna doesn't match with yours?

If you discover hulk on person A, you can see wich gene on person B is hulk if they have it

@Nabski89 
> What is genetic stability?
The stronger the power, the higher the genetic instability. These only occur when you give someone a mutation they don't have in their SE. It's a counterweight so you cant get all mutations without thought. I haven't decided yet on the effects of having a high instability

@Arianya Good idea, I'll check.

Time-Green committed:
removes unused css

swindly commented on code/datums/dna.dm:
```diff
@@ -171,6 +200,31 @@
 
+/datum/dna/proc/update_instability(alert=FALSE)
+	instability = 0
+	var/adjective = "still "
```
misleading variable name pls change

swindly commented on code/datums/mutations/actions.dm:
```diff
@@ -0,0 +1,65 @@
+/datum/mutation/human/telepathy
+	name = "Telepathy"
+	desc = "A rare mutation that allows the user to telephaticly communicate to others."
```
```suggestion
	desc = "A rare mutation that allows the user to telepathically communicate to others."
```

swindly commented on code/datums/mutations/actions.dm:
```diff
@@ -0,0 +1,65 @@
+		if(C.is_mouth_covered())
+			C.adjust_fire_stacks(2)
+			C.IgniteMob()
+			to_chat(C,"<span class='warning'>Something infront of your mouth caught fire!</span>")
```
```suggestion
			to_chat(C,"<span class='warning'>Something in front of your mouth caught fire!</span>")
```

nicbn commented:
I think it may be wise to rename hulk to avoid copyright problems

Nabski89 commented:
So let me make sure I understand this.

The way to do solo genetics (aka the only realistic way) is

Get a monkey, unlock and save as many mutations as you can on the console.
   Inject as many activators on yourself as you can.
   Inject mutators/injectors because haha fuck you you got unlucky this round.
Aquire another monkey to search through a second set of powers for things that you may have missed.
   Repeat injections.

How long do you think this is going to take a typical genetics player?
   Is there any consideration for the fact that goon rounds are typically longer than a /tg/ round?

anconfuzedrock commented:
Does the light eater really need to counter glowing though? Glowing sounds really niche and I don’t see why it shouldn’t counter a rare and already weak side antag

Time-Green commented:
@Nabski89 I'm not crazy fast, but those speed runners that do genetics in 10 minutes could probably still do that. I'd have to guess 30 minutes if you get some science upgrades halfway.

@Nabski89 I completely oversaw that. I'll add some form of gene sampling so you can still effectively do genetics if you happen to be alone

Time-Green committed:
Spellcheck by bobbahbrown and swindly

Time-Green committed:
Merge branch 'spicy-genes' of https://github.com/Time-Green/tgstation into spicy-genes

Time-Green commented:
It's less of an if, and more of a "they definitely will". Wich is intended

Barhandar commented:
Except that upgraded cloning is supposed to remove disabilities.

Fox-McCloud commented:
Why not go full Goon, with their UI and activation methodology? It presents the mini-game a lot better than this.

SouDescolado commented:
In case #40995 gets merged, make Ethereal spawn with the Glowy mutation, and if you can dna scan them, you can get it "for free"

Time-Green commented:
I didnt make any balance changes to cloning so then it wont. I just assumed it was already a thing.





> Why not go full Goon, with their UI and activation methodology? It presents the mini-game a lot better than this.

I really, really wanted to. But I couldn't. I don't have the skill in UI or patience to absolutely completely change it. Maybe one day.

Nabski89 commented:
You might want to change/remove the job specific gun for genetics, since you don't have injectors to shoot anymore.

I wouldn't say I'm a crazy speedrunner, but I've got 103 hours logged as geneticist right now. 30 minutes involving upgrades is super trash from a "getting to actually play the game and not just stand in a room playing the worst slots in the world".

One of the bigger balancing points currently is radiation. Without upgrades it's almost impossible to apply powers to yourself solo without requiring anti toxin for the rest of the shift. What's the planned balance on that.

Does this do anything to fix the secret monkey technique that allows you to find all the power blocks in 2 minutes?

Time-Green commented:
> You might want to change/remove the job specific gun for genetics, since you don't have injectors to shoot anymore.
> 
> I wouldn't say I'm a crazy speedrunner, but I've got 103 hours logged as geneticist right now. 30 minutes involving upgrades is super trash from a "getting to actually play the game and not just stand in a room playing the worst slots in the world".
> 
> One of the bigger balancing points currently is radiation. Without upgrades it's almost impossible to apply powers to yourself solo without requiring anti toxin for the rest of the shift. What's the planned balance on that.
> 
> Does this do anything to fix the secret monkey technique that allows you to find all the power blocks in 2 minutes?

There's a genetics gun? It will probably work with the new injectors and activators, assuming it wasnt shitcoded. I'll check.

About the time, it was a guess. You don't need to actually get every single mutation. The idea is to have it like goon: you get a few cool unique mutations and have fun with it. I still need to add more mutations for that to really get working. Stuff like midas touch, ice touch, matter eater and jumpy. 
https://wiki.ss13.co/Guide_to_Genetics Alot of the mutations will be coming from here, but not the crazy ones like wizard jaunt

The activotors and mutators do significantly less radiation damage, since they're single mutations. Furthermore, there's medication for that for a reason. 

Secret monkey technique? Is that what people abuse to cheese genetics? If I find out how it works I will code it out, if I havent already.

Nabski89 commented:
What's your name on discord, if you're in the tg discord I'll PM you about it there.

evsey9 commented:
If you don't port the radiation emitters that let you reroll DNA potential of a monkey/human then you're just gonna make genetics worse

Time-Green commented:
> If you don't port the radiation emitters that let you reroll DNA potential of a monkey/human then you're just gonna make genetics worse



> If you don't port the radiation emitters that let you reroll DNA potential of a monkey/human then you're just gonna make genetics worse

Maybe a researchable item for later, I certainly thought about it

Iamgoofball commented:
@Time-Green MSO is rolling out a tgui replacement soon, it'll work.

Iamgoofball commented:
also :+1: from me and I'll totally start porting over all goon mutations once this is ready

Fox-McCloud commented:
>I really, really wanted to. But I couldn't. I don't have the skill in UI or patience to absolutely completely change it. Maybe one day.

Ahh, so it's more that; I thought you were inherently opposed to it. Glad to hear this!

Iamgoofball commented:
@Nabski89 When goonetics was developed and implemented, the average goon round was 30-45 minutes. If anything, our rounds are longer than when goonetics was at it's prime, and they managed to make do.

@Time-Green Did you implement the gene sequencing minigame at least?

Iamgoofball commented:
For genetic instability, I'd say just go full goon on that, and make sure you port mutation reactions too.

evsey9 commented:
There are several key components Goon has that makes their genetics fun and interesting and not fuck-all boring. I think you're missing several of them.

1. Player interaction and chromosomes.
On Goon, you get filled DNA activators from activating genes in human mobs that have a player inside. They are very important as they give you materials and _chromosomes_. Chromosomes are very important in goonetics if you want to have a lot of powers at once, or want to have POWERFUL powers. 
2. As I've already said, radiation emitters.
Without radiation emitters, doing genetics will be a HUGE pain, because having to get a new monkey to get a different gene set is not fun at all.
3. Goonetics is SIMPLE.
The main part of goonetics is this: You get a monkey, you inject a gene activator into them which DOESNT require actually completing the sequencing minigame in order to activate it, you store the mutation and you make a DNA injector out of it. I do not know how it works currently with this PR, but I assume its the same thing with gene blocks of DAC and all that hexademical shit.
4. Fun activated powers.
Goon has alot of fun to use genetic powers that are active instead of passive, and this point also ties in with point 1 and the next point. Chromosomes let you heavily change how your active powers work, either making them safe to use, more powerful or having a lesser cooldown.
5. Genetic instability is more than a purely bad thing.
Genetic instability can have fun effects on goon too. While it DOES give you unwanted and potentially lethal effects after going below a certain point, it also gives activated powers a chance to fail and do something unexpected, "something unexpected" not always being a bad thing.

Iamgoofball commented:
^^^ yes make sure all of this is in

PKPenguin321 commented:
CRAZY HUGE 👍, but at the same time, **I dont think this system is worth having until we have all of the content that comes with it** like @evsey9 pointed out. I **REALLY** want this, but until we have it fully fleshed out and have a bunch more of the actually fun part (the powers being a big one), maybe it should wait.

evsey9 commented:
Make sure you have an ability that makes others make a sound and make sure that when your genetic stability is low it has a chance to play any sound from the game

Qustinnus commented:
this is a tmtime PR so i expect some godly bugs

Supermichael777 commented:
How about one that lets you slip restraints fast on a long cooldown. Hydrodynamic biology.

I have a few funny ideas that might be better as brain damage problems but they might work for this.

Time-Green commented:
> There are several key components Goon has that makes their genetics fun and interesting and not fuck-all boring. I think you're missing several of them.
> 
> 1. Player interaction and chromosomes.
>    On Goon, you get filled DNA activators from activating genes in human mobs that have a player inside. They are very important as they give you materials and _chromosomes_. Chromosomes are very important in goonetics if you want to have a lot of powers at once, or want to have POWERFUL powers.
> 2. As I've already said, radiation emitters.
>    Without radiation emitters, doing genetics will be a HUGE pain, because having to get a new monkey to get a different gene set is not fun at all.
> 3. Goonetics is SIMPLE.
>    The main part of goonetics is this: You get a monkey, you inject a gene activator into them which DOESNT require actually completing the sequencing minigame in order to activate it, you store the mutation and you make a DNA injector out of it. I do not know how it works currently with this PR, but I assume its the same thing with gene blocks of DAC and all that hexademical shit.
> 4. Fun activated powers.
>    Goon has alot of fun to use genetic powers that are active instead of passive, and this point also ties in with point 1 and the next point. Chromosomes let you heavily change how your active powers work, either making them safe to use, more powerful or having a lesser cooldown.
> 5. Genetic instability is more than a purely bad thing.
>    Genetic instability can have fun effects on goon too. While it DOES give you unwanted and potentially lethal effects after going below a certain point, it also gives activated powers a chance to fail and do something unexpected, "something unexpected" not always being a bad thing.

I was planning to add chromosomes and stats in a future PR, since it currently already got way too big. 
With enough time I could add the AT GC minigame, but it would get even bigger. If people don’t mind, then sure.

Also, genetic instability, can you give some examples of the effects? Goon doesn’t have a list of the effects and I only experienced statue and skelegib

Time-Green commented:
> CRAZY HUGE 👍, but at the same time, **I dont think this system is worth having until we have all of the content that comes with it** like @evsey9 pointed out. I **REALLY** want this, but until we have it fully fleshed out and have a bunch more of the actually fun part (the powers being a big one), maybe it should wait.

I really wanna do everything, but it’ll get way WAY too big for one PR. Most of it is on my todo list though

Iamgoofball commented:
Yes, get the framework and the minigame in first. We can add mutations in post.

optimumtact commented:
guarantee if you add the mutations from goon the players will complain until they're removed again and this will be a bigger flop than goofchem

but hey good luck with that.

also there's no reason to rename hulk ignore whoever said that.

Iamgoofball commented:
oranges is right, don't worry about adding new mutations with this PR tbh, it can be proposed in a follow up where someone like myself can make the PR and own all the loser skrubz who don't know how to debate on github properly

optimumtact commented:
you clearly didn't understand what I said at all

evsey9 commented:
@Time-Green take a look in giin2016 code by the way.
At least add chromosomes, rad emitters and genetic instability effects this PR.
Genetic stability, when low enough, gives your activated powers a chance to fail, when it's even lower it will have a big chance of making whatever you say indecipherable, and when it goes under something like 60 you will start getting toxin damage, losing your good mutations and getting bad mutations until your genetic stability is high enough.

Time-Green committed:
Converts structural enzymes into the new AT GC format

Now to go have fun and make this work with the dna console...

And yes it's not supposed to compile, I know

FantasticFwoosh commented:
Nice, i might be able to play genetics in the near future and not feel like a terrible person but instead a researcher when i obsessively collect all the traits on disks rather than give all assistants free hulk just because.

Can we add max limits to mutations upon a single person please? No combined dwarfism & hulk because mutation budgeting wouldn't allow it for instance.

- Also you can probably farm mutation effects from upgraded cloning, which will need to be looked at, **_sampler guns can probably be used mutually from the gene vault station objective_**, though it'd be nice if we could discover certain genes by analysing certain mobs.

EI: Slimes having voracious appetites, Gutlunches having effective digestive systems, sampling a gorilla gives enhanced muscle strength.

evsey9 commented:
dont add max mutation limits, there is genetic stability for that

Iamgoofball commented:
@optimumtact if you want us to stick with the same 7 mutations and add no more, don't bother with leading this guy on and close it now. it will be a waste of coder and maintainer time if we aren't utilizing the actual benefits of this state of design.

lolman360 commented:
people will always complain, it's change, and some people don't like change

anconfuzedrock commented:
Oh you goofy goober you, he said the goon mutations. More unique ones may not cause world war 3!

Supermichael777 commented:
@Time-Green https://wiki.ss13.co/Guide_to_Genetics#Possible_Mutations
This is goons list.
also gentics powers have a fail rate around 20% base on a single point of stability loss

Actually temporary stability damage could be funny, too unstable and you start randomly mutating hit zero and something horrible happens.

Time-Green commented:
Obviously need to improve the UI dramaticly and it's not done at all, but the beginning is there and it works.
![image](https://user-images.githubusercontent.com/7501474/47875652-61ba1800-de17-11e8-9833-b50c84772cfa.png)

Time-Green committed:
basic gene sequencing

for the love of god i am not done with this at all, so dont even bother reviewing it since I still need to adjust everything accordingly

Fox-McCloud commented:
Now that's starting to look more like what they have.

Time-Green committed:
Makes the UI really fucking good and makes everything functional again

optimumtact commented:
Wow all of you are missing the point, the point is our playerbase will blast the ass out of the goon mechanics and absolutely abuse the powers until they're all gone or nerfed.

That's why it will fail smh

subject217 commented:
somehow hulk is still around, it can't be that bad can it

evsey9 commented:
Can you give me an example of a goon power that could be abused by tg players, oranges?

evsey9 commented:
Also, if the players really DO abuse then it won't be the fault of the system - it will be the fault of the powers, which we can replace

nicbn commented:
it will be the fault of the players, which we can replace

Bawhoppen commented:
Players always will find a way to ruin everything nice we have.
That's why circuits had to go. A fun, unique, modular mechanic, full of options and creativity, yet was sadly driven into the ground by an onslaught of powergaming.

Very sad stuff.

yourdoom9898 commented:
@evsey9 One thing to mention, unless GoonGenetics has changed since I last touched it, you do not need another player to get filled DNA activators, you can instead humanize a money, give it genetic cancer by activating all of it's latent powers with activators, store any powers that you want to keep, then scramble it's DNA and start again, repeating until you have a billion research points and more chromosomes than you can really use up.

evsey9 commented:
@yourdoom9898 I know my goon shit, they changed it so that filled DNA activators gave you three chromosomes, an unlock thingy and mats each time BUT you only get filled activators from humans that have a player in them.

Time-Green committed:
Drasticly improves AI and makes it fully functional with everything

except disks, I still need to test that

Time-Green committed:
immersive d i s k functionality

anconfuzedrock commented:
inb4 laser eyes comes back

Time-Green commented:
I think I am gonna leave out laser eyes and wizard jaunt mutation, at least for now

tralezab commented:
if i make some mutations and pr them to your branch, would you consider adding them? I'm very excited for this pr, and genetics is one of my favorite jobs!

Time-Green commented:
> if i make some mutations and pr them to your branch, would you consider adding them? I'm very excited for this pr, and genetics is one of my favorite jobs!

Absolutely, just keep make sure they aren't too controversial and I'll probably merge it

davethwave commented:
There are only a few controversial mutations in goon so... I don't think we will have a problem with filling the void.

Time-Green committed:
Adds gene scanner

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into spicy-genes

davethwave commented:
Well, thanks too a good mentor/coworker. I am able to understand goon genetics. It's pretty easy and will help others understand too if this gets merged.

Arianya commented:
Did you ever get an answer on the glowy mutation and nightmares?

Time-Green commented:
> Did you ever get an answer on the glowy mutation and nightmares?

It's not intense enough to actually hurt the nightmare unless he's standing on you.

Arianya commented:
Alright, thanks for looking into it :+1:

Time-Green committed:
mutation stability

Time-Green committed:
Adds mutation combining

Time-Green commented:
Updated OP.

I am going to tweak some mutations and test the hell out of it

Time-Green commented:
Updated OP. Now also includes a very brief walkthrough

I am going to tweak some mutations and test the hell out of it

Time-Green committed:
Adds the ultimate anti cheese to mutagen and fixes more stuff

also (probably) breaks some stuff so I've got more to do for the next commit

Shadowflame909 commented:
Superfart is the most important genetics super power. If you don't add that it's a useless PR!

ExcessiveUseOfCobblestone commented:
Add the structure first with powers being discussed in their own PR tbh.

If you don't piecewise something like this then it simply won't get merged, especially since one maintainer already doesn't seem to have much interest.

anconfuzedrock commented:
Just to be sure, if monkey is still a human only mutation that’s prolly a really bad idea, taking into account the obvious and lesser form and blah
Ifthis gets merged I remember ideas for a ventcrawling mutation, I’d probably add some mutations like that and several arms and blatant ripoffs of caves of qud mutations and defects, like electro impulse and spontaneous combustion and night vision.

Time-Green commented:
> Just to be sure, if monkey is still a human only mutation that’s prolly a really bad idea, taking into account the obvious and lesser form and blah
> Ifthis gets merged I remember ideas for a ventcrawling mutation, I’d probably attempt to add some mutations like that and several arms and blatant ripoffs of caves of qud mutations and defects, like electro impulse and spontaneous combustion and night vision.

Everyone has monkey and hulk was moved to combining. 

I'm currently working on extendo arms and tralezeb PR'd void magnet mutation.
Apart from those, I'l justl add a few very simple mutations and wrap things up.

tralezab committed:
Void Magnet Gene (#1)

* that's a lotta testing! how about some more!!*

*no testing actually used

* fix up your broken code with flexseal!

* fixes and defines, flextape joke

* lowered the chances from 50 to 2. Nice.

Time-Green committed:
Update stats

Time-Green committed:
Merge branch 'spicy-genes' of https://github.com/Time-Green/tgstation into spicy-genes

Time-Green committed:
woops, fixes conflicts

Time-Green committed:
Stat balancing

Time-Green committed:
Fiery sweat

Time-Green committed:
Fixes, tweaks and QoL

Time-Green committed:
Makes dwarf default inert species

Time-Green committed:
Removes dwarf from normal gene pool

Time-Green committed:
Updates inert mutation handling

Time-Green committed:
Fixes a few bugs I found while reading

also fuck webedits

Militaires commented:
rebase to goon when

davethwave commented:
Can't wait until this gets test merged

Time-Green commented:
@davethwave am I supposed to ask or does it just kinda happen?

davethwave commented:
I have NO idea

Arianya commented:
Currently testmerged on Bagil

Arianya commented:
@Time-Green When people scan genetic sequences it sends global messages to everyone

![image](https://user-images.githubusercontent.com/23403236/48313057-163e0180-e5af-11e8-936d-6733f45d1a11.png)

Arianya commented:
Untestmerged until the above is fixed because hoo boy global messages

Arianya commented:
Additionally, a runtime:

```
[12:33:15] Runtime in dna.dm, line 143: list index out of bounds
proc name: create sequence (/proc/create_sequence)
usr: *no key*/(Peyton Leech)
src: null
call stack:
create sequence(/datum/mutation/human/race (/datum/mutation/human/race), 0, 8)
/datum/dna (/datum/dna): generate struc enzymes()
/datum/dna (/datum/dna): initialize dna(null)
Peyton Leech (/mob/living/carbon/human/dummy): Initialize(1)
Atoms (/datum/controller/subsystem/atoms): InitAtom(Peyton Leech (/mob/living/carbon/human/dummy), /list (/list))
Peyton Leech (/mob/living/carbon/human/dummy): New(1)
generate or wait for human dummy("dummy_preference_preview")
/datum/preferences (/datum/preferences): update preview icon()
/datum/preferences (/datum/preferences): ShowChoices(CrazyClown12 (/mob/dead/new_player))
CrazyClown12 (/mob/dead/new_player): Topic("src=\[mob_17];show_preferences...", /list (/list))
CrazyClown12 (/client): Topic("src=\[mob_17];show_preferences...", /list (/list), CrazyClown12 (/mob/dead/new_player))
CrazyClown12 (/client): Topic("src=\[mob_17];show_preferences...", /list (/list), CrazyClown12 (/mob/dead/new_player))```

Time-Green commented:
Oh boi thats really bad, I thought I removed all of those

Nickkoranda commented:
Pretty fun and engaging, a bit too engaging maybe-- I was stuck on the screen instead of comms. Maybe there could be ways to tone down some of the solving though? I had some cases where I had more than 64 possibilities, which I know I couldn't do without losing my sanity.

Time-Green committed:
Tweaks species mutations and fixes sequence scanners broadcasting globally

Time-Green commented:
> Pretty fun and engaging, a bit too engaging maybe-- I was stuck on the screen instead of comms. Maybe there could be ways to tone down some of the solving though? I had some cases where I had more than 64 possibilities, which I know I couldn't do without losing my sanity.

You're not supposed to guess all 64 sequences, unless you really want to I guess. The idea is to use the gene sequencer and check someone else to see if they have some of the unknown pairs there.

Time-Green commented:
@Arianya That was embarassing, fixed.

Arianya commented:
Retestmerged on Bagil

anconfuzedrock commented:
Fire sweat + cough should be how you get fire breath, like hulk being strength + radioactive.

Time-Green commented:
That’s a good combination, but I kinda want people to go out and catch a lizard

anconfuzedrock commented:
Right, but what probably should be changed
-if you lose the telepathy, you might keep the power
-humans can use firebreath activators (lizards can't use hulks at least)
-it sounds like geneticists haven't reported any negative effects of genetic instability. I haven't seen em either in my experience.

Time-Green commented:
> Right, but what probably should be changed
> -if you lose the telepathy, you might keep the power
> -humans can use firebreath activators (lizards can't use hulks at least)
> -it sounds like geneticists haven't reported any negative effects of genetic instability. I haven't seen em either in my experience.

Is the telepathy a suggestion or a bug?
Genetic stability worsens stuff like fire sweat and void magnet. It also has a meltdown when it hits zero. Right now there aren't that many powers that can get that special effect. I dont think anyone had a meltdown since there's not really that many powers yet

anconfuzedrock commented:
bug

adamkad1 commented:
on goons, low stability doesnt hurt, just increase chance of a power experiencing a critical failure, is it same here?

Time-Green commented:
> on goons, low stability doesnt hurt, just increase chance of a power experiencing a critical failure, is it same here?

Yeah, though the only ones affected are fire sweat and void magnet. Also it does hurt if you hit zero

adamkad1 commented:
race changing mutations when

Time-Green commented:
when someone codes it

adamkad1 commented:
right, so goon activators actually put letters in their places, is there a plan to do it here?

Time-Green commented:
> right, so goon activators actually put letters in their places, is there a plan to do it here?

I dont know what you saw but thats exactly what happens

Time-Green committed:
Fixes powers not going away and void being cast in void

also moves void magnet to actions.dm

kriskog commented:
I ran into some potentially unintended stuff. The mutations that give you spells don't remove the spells if you eat mutadone. This allows you to use void magnet without the drawback, and also to get the spells multiple times.

Also, I miss the option to reset sequences if I messed something up. ie if I try to guess multiple pairs and forget which were x-es.
I would also like to see different colored buttons or something for the values where you have made a change.

Time-Green commented:
I fixed the powers stuff already, the testmerge just needs to be updated.
I cant do the reset, but I can mark recent changes, thats a really good idea

Kawaii-Big-Boss commented:
This sounds like it could be pretty interesting. If it makes genetics more fun and less "rush hulk", then I'm down.

Ledwin1233 commented:
Can you add more combination genes? Half the fun is finding the weird and wacky combination genes on goon.

Time-Green commented:
I will definitely add more, but if I do it now there will be alot less mutations in the pool and there’s not super many yet

Time-Green commented on code/game/objects/items/devices/scanners.dm:
```diff
@@ -644,3 +645,65 @@ SLIME SCANNER
+	if(istype(O, /obj/machinery/computer/scan_consolenew))
+		var/obj/machinery/computer/scan_consolenew/C = O
+		if(C.stored_research)
+			to_chat(user, "<span class='notice'>[name] database updated.</span>")
```
I’ll fix it

ShizCalev commented:
#41453

Nabski89 commented:
What does upgrading the genetics machine parts do? I had expected it to give you more save slots or reduce your X-X pair chances, but it looks like it's still just radiation reduction. X-X mutations suck if you get more than one of them.

This is really tedious to play since it becomes "how quickly can you spam matching" compared to the previous click, wait 4-6 seconds, hope for good RNG. It's an easier to understand and explain system, but oh boy does it suck to work on.

The new mutations are fun however.

tralezab commented:
Very sorry about void bugs by the way.

Time-Green commented:
> Very sorry about void bugs by the way.

It was only one bug and pretty harmless

anconfuzedrock commented:
the double x needing another dude is another reason why the fire breath is so much of an asspain to get, I still think it should be a combo.

Time-Green commented:
I’ll add increasing storage and give people a very good reason to upgrade their stuff

subject217 commented on code/game/objects/items/devices/scanners.dm:
```diff
@@ -644,3 +645,65 @@ SLIME SCANNER
+	var/list/discovered = list() //hit a dna console to update the scanners database
+
+/obj/item/sequence_scanner/attack(mob/living/M, mob/living/carbon/human/user)
+	user.visible_message("<span class='notice'>[user] has analyzed [M]'s genetic seqequence.</span>")
```
it's "sequence".

Time-Green committed:
Adds joker algorithm

Time-Green committed:
fixes a typo

Time-Green commented:
DNA console can now guess a single X right every 20 minutes. This cooldown decreases with upgrades

optimumtact commented on code/__HELPERS/type2type.dm:
```diff
@@ -629,3 +629,9 @@
 		r += ascii2text(c)
 	return r
+
+/proc/string2list(str)
```
what does this actually do?

Qustinnus replied:
it turns a string into a list of characters 🤔

optimumtact replied:
yes but why, can you not treat strings as lists anyway?

Time-Green replied:
wait really?

anconfuzedrock commented:
not sure if intended, but you can easily find the race specific fire breath by simply scanning everyone at roundstart and finding which gene number the human crew lack (which can get the mutation from a mutator!)

AnturK reviewed (changes requested):
I don't have energy for detailed review but:
You're reinventing the wheel with body bound spells, that's what mob spell list is for.
If se modification is gone, se synchronization should be gone too, slapping new shit on it is bad idea.
There's a lot of back and forward references here not cleaned when needed now see runtimes and deletion failures in logs.

Time-Green commented:
I didnt know that was how spells worked, I’ll change that.
What do you mean with se synchronization? 
I’ll do some more private playtesting so I can see the logs, but I haven’t had any runtimes in a while.
I’ll double check everything and clean up sone code, going to take a while though.
Thanks for the review, I know it’s a pain

kriskog commented:
Is this what causes you to get disabilities after cloning, or did I miss another change? Seems to happen all the time with unupgraded cloners. Intended?

ShizCalev commented:
Happening during game load if a player has the character settings window open

```
[18:16:33] Runtime in dna.dm, line 139: Cannot read null.difficulty
proc name: create sequence (/proc/create_sequence)
usr: *no key*/(Marion Fryer)
src: null
call stack:
create sequence(/datum/mutation/human/race (/datum/mutation/human/race), 0, null)
/datum/dna (/datum/dna): generate struc enzymes()
/datum/dna (/datum/dna): initialize dna(null)
Marion Fryer (/mob/living/carbon/human/dummy): Initialize(1)
Atoms (/datum/controller/subsystem/atoms): InitAtom(Marion Fryer (/mob/living/carbon/human/dummy), /list (/list))
Marion Fryer (/mob/living/carbon/human/dummy): New(1)
generate or wait for human dummy("dummy_preference_preview")
/datum/preferences (/datum/preferences): update preview icon()
/datum/preferences (/datum/preferences): ShowChoices(DerpDirkins (/mob/dead/new_player))
DerpDirkins (/mob/dead/new_player): Topic("src=\[mob_11];show_preferences...", /list (/list))
DerpDirkins (/client): Topic("src=\[mob_11];show_preferences...", /list (/list), DerpDirkins (/mob/dead/new_player))
DerpDirkins (/client): Topic("src=\[mob_11];show_preferences...", /list (/list), DerpDirkins (/mob/dead/new_player))
```

Time-Green commented:
@ShizCalev thanks alot, I can work with that
@kriskog that’s always been a thing and I dont think I changed it. Only change in random mutation generation is that bad and neutral mutations share the same pool, so that might be what you’re experiencing.

kriskog commented:
@Time-Green 
Huh, weird. Is it possible this PR changes the frequency on bad mutations happening?

Time-Green commented:
> @Time-Green
> Huh, weird. Is it possible this PR changes the frequency of bad mutations happening? I have 700+ hours living, and I only noticed this for the first time yesterday where it happened twice in a row. Seems too much of a coincidence to me.

I checked and there’s always been a 30% chance for an unupgraded cloner to give a bad mutation

AnturK commented on code/controllers/subsystem/atoms.dm:
```diff
@@ -108,22 +108,25 @@ SUBSYSTEM_DEF(atoms)
+	var/list/mutations = subtypesof(/datum/mutation/human)
+	shuffle_inplace(mutations)
+	for(var/A in subtypesof(/datum/generecipe))
+		new A ()
```
There's absolutely no need to create and delete these with current code, use initial. (If you want more than two mutations in these you need to change this a bit)

AnturK replied:
No ?

Qustinnus replied:
wtf are you on about time

Time-Green replied:
wait no fuck that was just my experience because the loop needed to be typeless and then cast as the /datum/generecipe

AnturK commented on code/__DEFINES/DNA.dm:
```diff
@@ -56,7 +69,8 @@
 
-#define DNA_STRUC_ENZYMES_BLOCKS	18
+#define DNA_SEQUENCE_LENGTH			4
+#define DNA_STRUC_ENZYMES_BLOCKS	8
```
Rename to fit new se name.

AnturK commented on code/datums/dna.dm:
```diff
@@ -104,18 +108,47 @@
+	var/list/mutations_temp = GLOB.good_mutations + GLOB.bad_mutations + GLOB.not_good_mutations + bonus
+	mutation_index.Cut()
+	shuffle_inplace(mutations_temp)
+	if(ismonkey(holder))
```
This should delegated to mob function/property.

Time-Green replied:
It's already /datum/mutation/human

AnturK commented on code/datums/dna.dm:
```diff
@@ -342,26 +505,41 @@
-	var/datum/mutation/human/HM = pick(GLOB.good_mutations)
-	. = HM.force_give(src)
-
-/mob/living/carbon/proc/randmutvg()
```
very good helper disappeared ? Kinda unrelated but these helpers need to squashed into one + wrappers.

Time-Green replied:
It's just a persons native mutations now, so it'd be splitting up the few that they have even more.
I could also just make random mutations draw from the big pool. I'm not really sure wich is better

AnturK commented on code/datums/dna.dm:
```diff
@@ -386,7 +564,7 @@
 		for(var/i=1, i<=DNA_STRUC_ENZYMES_BLOCKS, i++)
 			if(prob(probability))
-				M.dna.struc_enzymes = setblock(M.dna.struc_enzymes, i, random_string(DNA_BLOCK_SIZE, GLOB.hex_characters))
+				M.dna.generate_struc_enzymes()
```
This only needs to be done once ?

Time-Green replied:
it resets all their dna blocks

Time-Green replied:
it also resets their dna blocks

AnturK commented on code/datums/footstep.dm:
```diff
@@ -1,99 +1,99 @@
-				playsound(T, pick(GLOB.barefootstep[T.barefootstep][1]),
-					GLOB.barefootstep[T.barefootstep][2] * v,
-					TRUE,
+/datum/component/footstep
```
Whitespace/merge issues

AnturK commented on code/datums/mutations.dm:
```diff
@@ -4,66 +4,54 @@ GLOBAL_LIST_EMPTY(mutations_list)
-	owner.dna.mutations.Add(src)
+	owner = H
+	dna = H.dna
+	dna.mutations[src] = type
```
What possible use this index has ? I don't see you using this anywhere and it doesn't make much sense.

AnturK commented on code/datums/mutations/radioactive.dm:
```diff
@@ -0,0 +1,18 @@
+
+/datum/mutation/human/radioactive/New()
+	..()
+	visual_indicators |= mutable_appearance('icons/effects/genetics.dmi', "radiation", -MUTATIONS_LAYER)
```
These probably should stay static.

AnturK commented on code/datums/mutations/sight.dm:
```diff
@@ -70,5 +75,5 @@
 
 /datum/mutation/human/laser_eyes/on_ranged_attack(mob/living/carbon/human/owner, atom/target, mouseparams)
-	if(owner.a_intent == INTENT_HARM)
+	if(owner.a_intent == INTENT_HELP)
```
???

Time-Green replied:
I thought it said HELP, thought it was retarted and changed it to HARM
i now realized i am the retarted

AnturK commented on code/datums/mutations.dm:
```diff
@@ -132,3 +129,15 @@ GLOBAL_LIST_EMPTY(mutations_list)
+
+
+/proc/get_initialized_mutation(B)
+	return GLOB.all_mutations[B]
```
These prototypes are bit messy, i feel a few statics would serve better here.

AnturK reviewed (commented):
More stuff, will continue later.

XDTM commented on code/datums/mutations/body.dm:
```diff
@@ -147,3 +160,57 @@
+	desc = "The user's muscles slightly expand."
+	quality = POSITIVE
+	text_gain_indication = "<span class='notice'>You feel strong.</span>"
+	difficulty = 16
```
does this do anything?

Iamgoofball replied:
Need to check but iirc on goon the effects for it are all on objects

Iamgoofball replied:
oh wait, read OP

XDTM replied:
oh i see, it's just for crafting hulk

XDTM commented on code/datums/mutations/body.dm:
```diff
@@ -147,3 +160,57 @@
+	desc = "The user's skin will randomly combust, but is generally alot more resilient to burning."
+	quality = NEGATIVE
+	text_gain_indication = "<span class='warning'>You feel hot.</span>"
+	text_lose_indication = "<span class'notice'>You feel alot cooler.</span>"
```
a lot

XDTM commented on code/datums/mutations/cold_resistance.dm:
```diff
@@ -1,11 +1,12 @@
 //Cold Resistance gives your entire body an orange halo, and makes you immune to the effects of vacuum and cold.
 /datum/mutation/human/cold_resistance
 	name = "Cold Resistance"
+	desc = "A strange mutation that makes the holder completely immune to hostile temperatures."
```
it's probably a good moment to rename it to Space Adaptation or something like that, since it also prevents damage from low pressure

XDTM commented on code/datums/mutations/sight.dm:
```diff
@@ -18,6 +19,7 @@
 //Blind makes you blind. Who knew?
 /datum/mutation/human/blind
 	name = "Blindness"
+	desc = "Completely inhibits peripheral brain function."
```
not particularly important, but i'd suggest making this about the eyes themselves since a brain issue would be the same as a brain trauma

evsey9 commented:
Can you get more than 8 mutations using DNA injectors/mutators or whatever they're called?

TheDreamweaver commented:
Yay! Finally someone accomplished it! Also now that race specific mutations are a thing, I have some neat ideas for the other races.

Time-Green committed:
Addresses reviews pt.1

Time-Green committed:
review pt2

Time-Green committed:
review pt 3

not done yet

Time-Green committed:
Fixes code part 4

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into spicy-genes

Time-Green commented:
> Can you get more than 8 mutations using DNA injectors/mutators or whatever they're called?

@evsey9 
Yeah, I said so in the OP. They'll make you unstable though

Time-Green committed:
removes unused define

Time-Green committed:
Possible fixes weird merge issue and adds more consistency

Time-Green committed:
Fixes character creation runtimes

Time-Green committed:
Turns randmutx into easy_randmut

also changes mutation_in_se to mutation_in_sequence

Time-Green commented:
Fixes roundstart runtimes, a bunch of inconsistencies and all the stuff that was pointed out. Except for the weird merge issue, I copied the file from master and it still just decides that it wants to remove and readd everything

Cyberboss commented:
Fix conflicts

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into spicy-genes

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into spicy-genes

Time-Green committed:
Fixes merge conflict by readding my icons

Time-Green committed:
Fixes dme

Time-Green commented:
@Cyberboss conflicts fixed

Time-Green committed:
Merge remote-tracking branch 'upstream/master' into spicy-genes

Vidiveni commented:
From testing live I can say this made genetics much more enjoyable. I run around scanning random people to figure out pairs.

adamkad1 commented:
say, shouldnt negative 'powers' give stability?

PKPenguin321 commented:
>say, shouldnt negative 'powers' give stability?

No, as a general rule
